migration/model.go

```diff
@@ -0,0 +1,135 @@
+package migration
+
+import (
+	"encoding/binary"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/orm"
+)
+
+var _ orm.CloneableData = (*Schema)(nil)
+
+func init() {
+	MustRegister(1, &Schema{}, NoModification)
+}
+
+func (s *Schema) Validate() error {
+	if err := s.Metadata.Validate(); err != nil {
+		return errors.Wrap(err, "metadata")
+	}
+	if s.Version < 1 {
+		return errors.Wrap(errors.ErrInvalidModel, "version must be greater than zero")
+	}
+	if s.Pkg == "" {
+		return errors.Wrap(errors.ErrInvalidModel, "pkg is required")
+	}
+	return nil
+}
+
+func (s *Schema) Copy() orm.CloneableData {
+	return &Schema{
+		Metadata: s.Metadata.Copy(),
+		Version:  s.Version,
+		Pkg:      s.Pkg,
+	}
+}
+
+// schemaID returns a detereministic ID of this schema instance. Created IDs
+// can be sorted using lexographical order from the lowest to the highest
+// version.
+func schemaID(pkg string, version uint32) []byte {
+	raw := make([]byte, len(pkg)+4)
+	copy(raw, pkg)
+	binary.BigEndian.PutUint32(raw[len(pkg):], version)
+	return raw
+}
+
+type SchemaBucket struct {
+	orm.Bucket
+}
+
+func NewSchemaBucket() *SchemaBucket {
+	b := orm.NewBucket("schema", orm.NewSimpleObj(nil, &Schema{}))
+	return &SchemaBucket{Bucket: b}
+}
+
+// MustInitPkg initialize schema versioninig for a given package name. This
+// registers a version one schema.
+// This function panics if not successful. It is safe to call this function
+// many times as duplicate registrations are ignored.
+func MustInitPkg(db weave.KVStore, packageName string) {
+	_, err := NewSchemaBucket().Create(db, &Schema{
+		Metadata: &weave.Metadata{Schema: 1},
+		Pkg:      packageName,
+		Version:  1,
+	})
+	// Duplicated initializations are ignored.
+	if err != nil && !errors.ErrDuplicate.Is(err) {
+		panic(err)
+	}
+}
+
+// CurrentSchema returns the current version of the schema for a given package.
+// It returns ErrNotFound if no schema version was registered for this package.
+// Minimum schema version is 1.
+func (b *SchemaBucket) CurrentSchema(db weave.ReadOnlyKVStore, packageName string) (uint32, error) {
```

Comment:
Please add a test for this when there are many schemas, and I will fix this code to use iterator

Reply:
Done in https://github.com/iov-one/weave/pull/547/commits/5ff4d8bf11f76ffef37bc0c66030d198b16ce8b5